src/SFA.DAS.EmployerCommitmentsV2.Web/Mappers/Cohort/DetailsViewModelMapper.cs

```diff
@@ -38,37 +42,85 @@ public async Task<DetailsViewModel> Map(DetailsRequest source)
                 LegalEntityName = cohort.LegalEntityName,
                 ProviderName = cohort.ProviderName,
                 Message = cohort.LatestMessageCreatedByProvider,
-                Courses = GroupCourses(draftApprenticeships)
+                Courses = await GroupCourses(draftApprenticeships.ToList())
             };
         }
 
-        private IReadOnlyCollection<DetailsViewCourseGroupingModel> GroupCourses(IEnumerable<DraftApprenticeshipDto> draftApprenticeships)
+        private async Task<IReadOnlyCollection<DetailsViewCourseGroupingModel>> GroupCourses(IList<DraftApprenticeshipDto> draftApprenticeships)
         {
-            return draftApprenticeships
-                .GroupBy(a => new {a.CourseCode, a.CourseName})
+            var apprenticeFundingBandCaps = await Task.WhenAll(draftApprenticeships.Select(async a => new { a.Id, FundingBandCap = await GetFundingBandCap(a.CourseCode, a.StartDate) }));
```

Comment:
This operation is going to be CPU bound (not I/O) so is not going to benefit from running async. Since this is going to call the funding csap method for all the draft apprenticeships anyway would the code be simpler if the call to get GetFundingBandCap was moved to the L70 and the join replaced with a select?

Reply:
unfortunately because GetFundingCap uses an async call, it was easier to execute it this way, otherwise the code became complex.